src/test/spec/unified_runner/test_file.rs

```diff
@@ -0,0 +1,379 @@
+
+use crate::{
+    bson::{doc, Bson, Deserializer as BsonDeserializer, Document},
+    concern::{Acknowledgment, ReadConcernLevel},
+    error::Error,
+    options::{
+        ClientOptions,
+        CollectionOptions,
+        DatabaseOptions,
+        HedgedReadOptions,
+        ReadConcern,
+        ReadPreference,
+        SelectionCriteria,
+        WriteConcern,
+    },
+    test::{TestClient, DEFAULT_URI},
+};
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct TestFile {
+    pub description: String,
+    #[serde(deserialize_with = "deserialize_schema_version")]
+    pub schema_version: Version,
+    pub run_on_requirements: Option<Vec<RunOnRequirement>>,
+    pub allow_multiple_mongoses: Option<bool>,
+    pub create_entities: Option<Vec<Entity>>,
+    pub initial_data: Option<Vec<CollectionData>>,
+    pub tests: Vec<TestCase>,
+}
+
+fn deserialize_schema_version<'de, D>(deserializer: D) -> std::result::Result<Version, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let mut schema_version = String::deserialize(deserializer)?;
+    // If the schema version does not contain a minor or patch version, append as necessary to
+    // ensure the String parses correctly into a semver::Version.
+    let count = schema_version.split('.').count();
+    if count == 1 {
+        schema_version.push_str(".0.0");
+    } else if count == 2 {
+        schema_version.push_str(".0");
+    }
+    Version::parse(&schema_version).map_err(|e| serde::de::Error::custom(format!("{}", e)))
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct RunOnRequirement {
+    min_server_version: Option<String>,
+    max_server_version: Option<String>,
+    topologies: Option<Vec<Topology>>,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "lowercase", deny_unknown_fields)]
+pub enum Topology {
+    Single,
+    ReplicaSet,
+    Sharded,
+    #[serde(rename = "sharded-replicaset")]
+    ShardedReplicaSet,
+}
+
+impl RunOnRequirement {
+    pub async fn can_run_on(&self, client: &TestClient) -> bool {
+        if let Some(ref min_version) = self.min_server_version {
+            let req = VersionReq::parse(&format!(">= {}", &min_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref max_version) = self.max_server_version {
+            let req = VersionReq::parse(&format!("<= {}", &max_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref topologies) = self.topologies {
+            if !topologies.contains(&client.topology().await) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub enum Entity {
```

Comment:
It's a little confusing that there are two things called `Entity`. Could they be named something more specific to make it clear which is which, e.g. `EntityEntry` for the one stored in the map?